packages/@aws-cdk/aws-elasticsearch/lib/domain.ts

```diff
@@ -0,0 +1,611 @@
+
+  /**
+   * Metric for KMS key being inaccessible.
+   *
+   * @default maximum over a minute
+   */
+  metricKMSKeyInaccessible(clientId: string, props?: MetricOptions): Metric;
+
+  /**
+   * Metric for number of searchable documents.
+   *
+   * @default maximum over a minute
+   */
+  metricSearchableDocuments(clientId: string, props?: MetricOptions): Metric;
+
+  /**
+   * Metric for search latency.
+   *
+   * @default maximum over a minute
+   */
+  metricSearchLatency(clientId: string, props?: MetricOptions): Metric;
+
+  /**
+   * Metric for indexing latency.
+   *
+   * @default maximum over a minute
+   */
+  metricIndexingLatency(clientId: string, props?: MetricOptions): Metric;
+}
+export class Domain extends cdk.Resource implements IDomain {
+  /**
+   * @attribute
+   */
+  public readonly domainArn: string;
+
+  /**
+   * @attribute
+   */
+  public readonly domainName: string;
+
+  /**
+   * @attribute
+   */
+  public readonly domainEndpoint: string;
+
+
+  private readonly domain: CfnDomain;
+
+  private readonly slowSearchLogGroup?: logs.LogGroup;
+
+  private readonly slowIndexLogGroup?: logs.LogGroup;
+
+  private readonly appLogGroup?: logs.LogGroup;
+
+  constructor(scope: cdk.Construct, id: string, props: DomainProps) {
+    super(scope, id, {
+      physicalName: props.domainName,
+    });
+
+    // Setup logging
+    const logGroups: logs.LogGroup[] = [];
+
+    if (props.logPublishingOptions?.slowSearchLogGroup) {
+      this.slowSearchLogGroup = props.logPublishingOptions.slowSearchLogGroup;
+      logGroups.push(this.slowSearchLogGroup);
+    } else if (props.logPublishingOptions?.slowSearchLogEnabed) {
+      this.slowSearchLogGroup = new logs.LogGroup(this, 'SlowSearchLogs', {
+        logGroupName: `elasticsearch/domains/${props.domainName ?? id}/slow-search-logs`,
+        retention: logs.RetentionDays.ONE_MONTH,
+      });
+      logGroups.push(this.slowSearchLogGroup);
+    }
+
+    if (props.logPublishingOptions?.slowIndexLogGroup) {
+      this.slowIndexLogGroup = props.logPublishingOptions.slowIndexLogGroup;
+      logGroups.push(this.slowIndexLogGroup);
+    } else if (props.logPublishingOptions?.slowIndexLogEnabed) {
+      this.slowIndexLogGroup = new logs.LogGroup(this, 'SlowIndexLogs', {
+        logGroupName: `elasticsearch/domains/${props.domainName ?? id}/slow-index-logs`,
+        retention: logs.RetentionDays.ONE_MONTH,
+      });
+      logGroups.push(this.slowIndexLogGroup);
+    }
+
+    if (props.logPublishingOptions?.appLogGroup) {
+      this.appLogGroup = props.logPublishingOptions.appLogGroup;
+      logGroups.push(this.appLogGroup);
+    } else if (props.logPublishingOptions?.appLogEnabed) {
+      this.appLogGroup = new logs.LogGroup(this, 'AppLogs', {
+        logGroupName: `elasticsearch/domains/${props.domainName ?? id}/application-logs`,
+        retention: logs.RetentionDays.ONE_MONTH,
+      });
+      logGroups.push(this.appLogGroup);
+    }
```

Comment:
Please try to refactor this as it contains a lot of very similar code. 

Also, I don't think we should use the log group provided by the user if he/she also disabled logging, which you only check later...To avoid confusion lets add a validation that if a log group is given, the logging option must be enabled.

Reply:
Good point. I have refactored it and I now only configure the log groups if they are enabled.

Reply:
I think I should change the log group default setting to true though so that if someone specified a log group they don't also have to set it to enabled.

Reply:
But this would mean that if a log group isn't specified, it will be created by default. I'm actually ok with this, enabling logs is a good practice. However, if this incurs additional charges, we should reconsider. Please check?

Reply:
I have refactored it. The log groups are only configured if they are enabled and by default they are disabled.